### PR TITLE
add pentree example and data

### PR DESCRIPTION
**Pull request by skaae**
Penntree example

Moved from https://github.com/craffel/nntools/pull/33

### CODE REVIEW

**Comment by craffel** [comment]
This is awesome, but I think it might be over the top as an example and may be more appropriate as a separate (from the main lasagne codebase) example script.  I'm not sure what @benanne and @f0k's expectations as far as the examples are... I think the current examples (with the fake time-delay data) are too simple, but I don't know of a "MNIST for recurrence"; it seems like the best example is one which shows how to construct the network with minimal other code.  I think it would be best to not distribute data in the github repo, and instead use a downloader script as with MNIST.  One (sort of dumb) example we could do is MNIST with RNNs; it's been done before...

File: examples/pentree.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+#    embedding --> LSTM1 --> LSTM2 --> output network --> predictions
+l_inp = lasagne.layers.InputLayer((BATCH_SIZE, MODEL_SEQ_LEN))
+
+l_emb = lasagne.layers.EmbeddingLayer(
+    l_inp,
+    input_size=vocab_size,       # size of embedding = number of words
+    output_size=embedding_size,  # vector size used to represent each word
+    W=INI)
+
+if dropout_frac > 0:
```
Comment (benanne) [inline examples/pentree.py]: this shouldn't be necessary, a dropout layer is just a pass through if the dropout fraction is 0, so it doesn't add any overhead.

**Comment by benanne** [comment]
It looks pretty beefy for an example, but looking closer it really isn't all that much more complicated than the current MNIST example. I feel it may be a lot easier to digest in the form of an IPython notebook though.

**Comment by craffel** [comment]
> I feel it may be a lot easier to digest in the form of an IPython notebook though.

Yes, although I'd argue the same is true for the MNIST example(s).  Now that github renders ipynb itself, I might advocate for going that route... https://github.com/craffel/Lasagne-tutorial/blob/master/examples/tutorial.ipynb

**Comment by f0k** [comment]
> Yes, although I'd argue the same is true for the MNIST example(s). Now that github renders ipynb itself, I might advocate for going that route...

A notebook would be good for a tutorial, but we should still have a plain `.py` example file in Lasagne that people can copy and adapt. Regarding the pentree example, we could put that into the Lasagne/Recipes repository. I'd be fine with a feedforward net on MNIST being the only example included in the Lasagne repository, just for the sake of having an example shipping with the library, and all other examples living in Recipes.

**Comment by craffel** [comment]
> I'd be fine with a feedforward net on MNIST being the only example included in the Lasagne repository, just for the sake of having an example shipping with the library, and all other examples living in Recipes.

Sounds good to me.  In that case though, it's probably worth explaining more carefully in the recurrent.py docstring how to mix recurrent layers and dense layers.

**Comment by skaae** [comment]
It's fine by me if it goes in the recipes. I do think that the example is nice in a more general way because it shows how you do a language model with all the fancy batch reordering and also how you calculate perplexity. Both things was non-obvious (at least to me) and not many examples exists. 

I think its important to have some non-trivial examples. I guess new users will use the framework which has an example closest to their needs. 

I don't have any preference for notebook vs py files.

**Comment by f0k** [comment]
> I don't have any preference for notebook vs py files.

The deeplearning tutorials usually have both: A tutorial-like text interspersed with code walking you through an implementation, and a download link to a .py file you can use directly. We might want to do the same for some of the examples (but probably not for all of them).

**Comment by benanne** [comment]
Sounds like a lot of work :) But yeah, it would be nice to have a notebook version and a python file version of at least some of them.

**Comment by craffel** [comment]
Closing this as it sounds like it will be a recipe, or at least an ipynb.  We can take care of it after this fork is merged.

**Pull request closed without merging**